### PR TITLE
Completed lru cache implementation in javascript

### PR DESCRIPTION
**Pull request by erixbiz**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [ ] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

---
Completed the write-up on LRU Cache Implementation in Javascript.

### CODE REVIEW

**Comment by geoffrey45** [comment]
@erixbiz Thank you for submitting your article for review.

Carefully go through the text that I have added to this pull request's header, checking the necessary check boxes.

You can check a box by adding an `x` in the `[ ]` like this: `[x]`

---
Another thing, you have added your code as images. Please use code blocks instead.

File: content/articles/lru-cache-implementation-in-javascript/index.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
+In-memory caches such as Memcached and Redis are key-value stores between your application and your data storage. Since the data is held in RAM, it is much faster than typical databases where data is stored on a disk. RAM is more limited than disk. Cache invalidation algorithms such as least recently used (LRU) can help invalidate 'cold' entries and keep 'hot' data in RAM.
+
```
Comment (geoffrey45) [inline content/articles/lru-cache-implementation-in-javascript/index.md:2]: Please explain what is `'cold' entries` and `'hot' data`

File: content/articles/lru-cache-implementation-in-javascript/index.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+We can make optimize the shifting by eliminating it. To do so we use a double liked list. Why?
+
+Still using the example 1,2,3,4. We needed to add 8. Since it wasn't in our list, what we did was remove the oldest element 4 and shift the entire list items. In a doubly-linked list, all you need to do is to remove the rear and update the rear element to the previous element. Then add a new node and make it point to the second element and update the front. The entire operation is 0(1) time complexity.
+
+We can also optimize searching. How?
+
+Still using our previous example of cache size 4, with 1,2,3,4. searching for 8. We can optimize our search by using a hash table. With a Hash, you will have a key and a value pair. The Key in this case will be the actual values in our cache. The value will be the address of our value as shown in the diagram below.
+
+[Hash Table & Double Linked List Diagram](https://leetcode.com/problems/ddl_hash.png/)
+
```
Comment (geoffrey45) [inline content/articles/lru-cache-implementation-in-javascript/index.md:37]: Please download the image and add a link to the source. This way, if the source fails, you still have the image on your article.

**Review by geoffrey45 — CHANGES_REQUESTED** [review]
@erixbiz Please check the comments below.

**Comment by WanjaMIKE** [comment]
Plagiarism check done.

@erixbiz could you correct the following paragraph which appeared to have been copied verbatim.

```
In-memory caches such as Memcached and Redis are key-value stores between your application and your data storage. Since the data is held in RAM, it is much faster than typical databases where data is stored on a disk. RAM is more limited than disk. Cache invalidation algorithms such as least recently used (LRU) can help invalidate 'cold' entries and keep 'hot' data in RAM.
```

**Comment by erixbiz** [comment]
@geoffrey45 I have made all the changes requested and gone through it against the checklist as well. Looking forward to hearing from you once reviewed. Thank you.

**Comment by geoffrey45** [comment]
Hello @erixbiz Looks like you have 5k+ files changed. This was caused by force pushing your last commit.

Please download your latest files and open a new pull request.

Leave a comment for me in your new pull request.

**Comment by geoffrey45** [comment]
@erixbiz Please let me know when you open the new pull request.

**Pull request closed without merging**